pkgs/top-level/config.nix

```diff
@@ -0,0 +1,204 @@
+# This file defines the structure of the `config` nixpkgs option.
+
+{ lib, localSystem, crossSystem, config, ... }:
+
+with lib;
+
+let
+
+  mkMeta = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = args.description or ''
+      Whether to ${args.feature} while evaluating nixpkgs.
+    '' + ''
+      Changing the default will not cause any rebuilds.
+    '';
+  });
+
+  mkMassRebuild = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = (args.description or ''
+      Whether to ${args.feature} while building nixpkgs packages.
+    '') + ''
+      Changing the default will cause a mass rebuild.
+    '';
+  });
+
+  mkUse = args: mkMassRebuild (args // {
+    description = (args.description or ''
+      Whether to enable ${args.feature} support for nixpkgs packages.
+    '');
+  });
+
+  # Flipped optionalAttrs variants for convenience
+  whenEnabled = args: value: lib.optionalAttrs value args;
+  whenDisabled = args: value: lib.optionalAttrs (!value) args;
+
+  use = config.use;
+
+in {
+
+  options = {
+
+    /* Internal stuff */
+
+    warnings = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      internal = true;
+    };
+
+    /* Config options */
+
+    allowUnfree = mkMeta {
+      description = "Whether to allow evaluation of unfree packages.";
+    };
+
+    checkMeta = mkMeta {
+      feature = "check <literal>meta</literal> attributes of all the packages";
+    };
+
+    doCheckByDefault = mkMassRebuild {
+      feature = "run <literal>checkPhase</literal> by default";
+    };
+
+    /* Nixpkgs's equivalent of Gentoo's global USE flags (https://gentoo.org/support/use-flags/).
+     *
+     * To users. Changing any of these to a non-default value will void your
+     * warranty! That is, at the very least, you will not be able to benefit
+     * from Hydra binary cache, it is also entirely possible (likely, even) that
+     * your system will not even build (especially if you change more than one
+     * flag at the same time) as most packages are not tested in non-default
+     * configurations.
+     *
+     * To developers. When adding a new flag here consult the link above,
+     * everything that has a Gentoo eqivalent should probably use the same name.
+     */
+    use = {
+
+      headless = mkUse {
+        default = false;
+        description = "Whether to build nixpkgs packages for a headless system.";
+      };
+
+      /* Graphical toolkits */
+
+      noGraphics = mkUse {
+        default = use.headless;
+        description = "Whether to build nixpkgs without support for any graphical toolkits.";
+      };
+
+      X = mkUse {
+        feature = "X11";
```

Comment:
It's a bit weird to have one `feature` attribute getting put into `Whether to enable ${args.feature} support for nixpkgs packages.`, accepts values like "X11", and another `feature` attribute getting put into `Whether to ${args.feature} while evaluating nixpkgs.`, accepting values like "run foobar". It might make sense to give them a different name, the latter one could be renamed to `action`.

In addition I fear it might end up like `mkEnableOption "foo"` where people don't know that `Whether to enable ${arg}.` gets put around their argument. But I guess writing such config options isn't as common as NixOS modules.